File: Leetcode/sbgo/Leetcode993-二叉树的堂兄弟节点.go
package sbgo

func isCousins(root *TreeNode, x int, y int) bool {
	var xp, yp *TreeNode
	var xd, yd int
	var fdx, fdy bool
	check := func(node, parent *TreeNode, deep int) {
		if node.Val == x {
			xp = parent
			xd = deep
			fdx = true
		}
		if node.Val == y {
			yp, yd, fdy = parent, deep, true
		}
	}
	type pair struct {
		node *TreeNode
		deep int
	}
	que := []pair{{root, 0}}
	check(root, nil, 0)
	for len(que) > 0 {
		node, deep := que[0].node, que[0].deep
		que = que[1:]
		if node.Left != nil {
			que = append(que, pair{node.Left, deep + 1})
			check(node.Left, node, deep+1)
		}
		if node.Right != nil {
			que = append(que, pair{node.Right, deep + 1})
			check(node.Right, node, deep+1)
		}
		if fdx && fdy {
			break
		}
	}
	return xd == yd && xp != yp
}

File: Leetcode/PY/Leetcode993-二叉树的堂兄弟节点.py
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        xp, xd, fdx = None, None, False
        yp, yd, fdy =  None, None, False
        def check(node,parent,deep):
            if node.val == x:
                nonlocal xp, xd, fdx
                xp,xd,fdx = parent,deep,True
            if node.val == y:
                nonlocal yp, yd, fdy    
                yp, yd, fdy = parent,deep,True
        q = []
        q.append((root,0))
        while q:
            node,deep = q.pop(0)
            if node.left:
                q.append((node.left,deep + 1))
                check(node.left,node,deep+1)
            if node.right:
                q.append((node.right,deep+1))
                check(node.right,node,deep+1)
            if fdx and fdy:
                break
        return xd == yd and xp != yp